mlir/lib/ExecutionEngine/ROCm/BackendUtils.cpp

```diff
@@ -39,19 +39,106 @@
 // TODO: remove this once the rocm_agent_enumerator is ready
 #include "hip/hip_runtime.h"
 
+#include <cstring>
 #include <mutex>
 #include <numeric>
 
 using namespace mlir;
 
 static constexpr const char kRunnerProgram[] = "mlir-rocm-runner";
 static constexpr const char kRocmAgentEnumerator[] = "rocm_agent_enumerator";
+static constexpr const char gcnArchDelimiter[] = ":";
+static constexpr const char targetTriple[] = "amdgcn-amd-amdhsa";
+
+namespace {
+// TODO: remove this once the rocm_agent_enumerator is ready
+void getGpuGCNArchName(hipDevice_t device, std::string &gcnArchName) {
+  hipDeviceProp_t props;
+  hipError_t result = hipGetDeviceProperties(&props, device);
+  if (result != hipSuccess) {
+    gcnArchName = "";
+    return;
+  }
+
+  const char *pArchName = props.gcnArchName;
+  gcnArchName.assign(pArchName);
+}
+
+// This function convert the arch name to target features:
+// gfx908:sramecc+:xnack- to +sramecc,-xnack
+std::string getFeaturesFromArchName(const std::string &gcnArchName) {
+  std::string features;
+
+  // First step: get rid of the arch name in feature string
+  std::size_t firstSeperatorLoc = gcnArchName.find(gcnArchDelimiter);
+  if (firstSeperatorLoc == std::string::npos) {
+    return features;
+  }
+  std::string gcnArchFeature = gcnArchName.substr(firstSeperatorLoc);
+
+  // Second step: put each feature name to the vector
+  std::string token;
+  std::vector<std::string> featureTokens;
+  auto convertFeatureToken = [](std::string &token) {
+    if (token.back() != '+' && token.back() != '-') {
+      llvm::errs() << "malformed token: " << token << ", must end with +/-.\n";
+      return token;
+    }
+    token.insert(token.begin(), token.back());
+    token.pop_back();
+    return token;
+  };
+
+  size_t len = strlen(gcnArchDelimiter);
+  size_t featureStart = 0;
+  size_t featureEnd = 0;
+  for (size_t found = 0;
+       (found = gcnArchFeature.find(gcnArchDelimiter, found)) !=
+       std::string::npos;
+       ++found) {
+    featureStart = found;
+    featureEnd = gcnArchFeature.find(gcnArchDelimiter, found + 1) - len;
```

Comment:
You're subtracting `len` here but not on `featureStart` - you sure that's correct?

Reply:
A couple of lines below when the substring is to be obtained from the start/end, I added that offset there. But you are right, it should better off initialize with the correct `featureStart`, will address.

Reply:
Addressed in 42f7e7c